lib/IRGen/GenMeta.cpp

```diff
@@ -334,6 +334,33 @@ void IRGenModule::emitNonoverriddenMethodDescriptor(const SILVTable *VTable,
   getAddrOfLLVMVariable(entity, init, DebugTypeInfo());
 }
 
+void IRGenModule::addVTableTypeMetadata(llvm::GlobalVariable *var,
+    SmallVector<std::pair<Size, SILDeclRef>, 8> vtableEntries) {
+  if (!getOptions().VirtualFunctionElimination)
+    return;
+
+  for (auto ventry : vtableEntries) {
+    auto method = ventry.second;
+    auto offset = ventry.first.getValue();
+    var->addTypeMetadata(offset, typeIdForMethod(*this, method));
+  }
+
+  switch (getOptions().VCallVisibility) {
```

Comment:
Why not deriving the VCall visibility from the method visibility?
Setting the visibility for _all_ calls sounds dangerous, e.g. if a public method gets a "File" visibility.

Reply:
You can call a private/internal method from a function, annotated with `@_optimize(none)`. That will keep the virtual call alive

Reply:
Changed the PR to infer the vtable visibility from the class decl's access scope visibility.